openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md

```diff
@@ -0,0 +1,843 @@
+- **GIVEN** multiple cred_store plugins are registered
+- **WHEN** the gateway resolves its plugin
+- **THEN** it chooses the plugin with the lowest priority
+
+#### Scenario: Scoped client resolution
+- **GIVEN** the gateway selected plugin instance id `gts.x.core.modkit.plugin.v1~x.core.cred_store.plugin.v1~vendor.pkg.plugin.v1`
+- **WHEN** the gateway requests a client via `ClientHub.get_scoped::<dyn CredStorePluginApi>(ClientScope::gts_id(&plugin_instance_id))`
+- **THEN** the plugin client is resolved and used to serve the request
+
+### Requirement: cred_store plugin persistence, access control, and cryptography
+The system SHALL define how `cred_store` plugins persist secrets, enforce access control, and apply cryptography.
+
+The gateway SHALL be responsible for:
+- Performing request authentication and high-level authorization (role/permission checks) before calling the plugin.
+- Passing request-scoped `SecurityCtx` into the plugin.
+
+Each plugin SHALL be responsible for:
+- Enforcing tenant/user scoping when it interacts with any persistent store.
+- Ensuring that secret material is never logged.
+
+#### Access control (DB-backed plugins)
+For DB-backed plugins, access control SHALL be enforced using the `modkit-db` secure ORM layer.
+
+DB-backed plugins SHALL:
+- Use a secure DB access wrapper (e.g., `SecureConn`) for all queries.
+- Derive an `AccessScope` from the request `SecurityCtx` and apply it to all queries.
+- Use scoped CRUD operations that automatically produce deny-by-default behavior for empty scopes.
+
+#### Scenario: DB-backed plugin denies cross-tenant reads
+- **GIVEN** a secret record exists for tenant A with secret id `S`
+- **WHEN** a caller from tenant B calls `CredStorePluginApi::get_secret_material(ctx, S, secret_type_id)`
+- **THEN** the plugin query is scoped to tenant B
+- **AND** the plugin returns a not-found or permission-denied error
+
+#### Cryptography (DB-backed plugins)
+DB-backed plugins SHALL encrypt secret material before persisting it.
+
+DB-backed plugins SHOULD implement envelope encryption:
+- A per-record Data Encryption Key (DEK) encrypts the secret plaintext
+- A Key Encryption Key (KEK) encrypts (wraps) the DEK
+
+The system SHOULD support the following AEAD algorithms for DEK encryption:
+- `AES-256-GCM`
+- `XChaCha20-Poly1305`
+
+The plugin SHOULD bind context into Additional Authenticated Data (AAD), including:
+- `tenant_id`
+- `client_id` (if present)
+- `secret_id`
+- `secret_type_id`
+
+#### Scenario: Envelope encryption
+- **WHEN** the plugin persists secret material
+- **THEN** the plugin encrypts the plaintext with a per-record DEK using an AEAD cipher
+- **AND** the plugin stores only ciphertext and encryption metadata
+- **AND** the plugin stores the DEK only in wrapped form (encrypted by KEK)
+
+### Requirement: cred_store SQLite-backed plugin
+The system SHALL support a `cred_store` plugin implementation that persists secrets in a local database (SQLite/SeaORM).
```

Comment:
probably for sqlite we should use this https://www.sqlite.org/see/doc/trunk/www/readme.wiki